numpy/core/src/multiarray/arrayobject.c

```diff
@@ -462,6 +451,27 @@ PyArray_ResolveWritebackIfCopy(PyArrayObject * self)
 
 /*********************** end C-API functions **********************/
 
+
+/* dealloc must not raise an error, best effort try to write
+   to stderr and clear the error
+*/
+
+static NPY_INLINE void
+WARN_IN_DEALLOC(PyObject* warning, const char * msg) {
+    if (PyErr_WarnEx(warning, msg, 1) < 0) {
+        PyObject * s;
+        s = PyUString_FromString(msg);
+        if (s) {
+            PyErr_WriteUnraisable(s);
+            Py_DECREF(s);
+        }
+        else {
+            PyErr_WriteUnraisable(Py_None);
+        }
+    }
+};
+#undef MSG_FromString
```

Comment:
I may have missed something, but I could not find the reason for this line anywhere.

Reply:
This pull request allows completing the confusing part of pull request #9639, and now any call to ``PyArray_SetUpdateIfCopyBase`` will emit a ``DeprecationWarning``, also any time ``array_dealloc`` finds an ndarray with ``WRITEBACKIFCOPY`` or ``UPDATEIFCOPY`` semantics it will emit a ``DeprecationWarning``. See also the change in the numpy API documentation. 

It can only be done once this pull request lands since otherwise any call to ``nditer`` that generates an operand with ``WRITEBACKIFCOPY`` semantics would have raised this warning. After this pull request we can and should warn, since we provide an alternative mechanism of using a context. Using a context prevents reaching ``array_dealloc`` without resolving the ``WRITEBACKIFCOPY`` semantics

Reply:
Sorry, that was me being stupid. But IMO it's better not to alias functions with the preprocessor like this (see scipy/scipy#8010). I don't think you're even saving any lines.

Reply:
By "this" do you mean ``WARN_IN_DEALLOC`` (which is a function, not a preprocessor macro, and used twice so it saves ~10 lines) or ``PyUString_FromString`` (which I did not invent)?

Reply:
or MSG_FromString

Reply:
What I mean is that you should do this:

```
+#if PY_MAJOR_VERSION < 3
+        s = PyString_FromString(msg);
+#else
+        s = PyUnicode_FromString(msg);
+#endif
```

What's absolutely horrendous is that these changes show up in the raw patch but not the GitHub web view, which is why I missed them the first time. That is somewhat disturbing.

Reply:
![image](https://user-images.githubusercontent.com/15351036/33235591-92399bfa-d200-11e7-846e-ac4604cb72a1.png)
![image](https://user-images.githubusercontent.com/15351036/33235592-994518c0-d200-11e7-9364-8490b1cd6329.png)

Reply:
fixed in 8b553d0